### ISSUE TITLE
Use delegated constructors to remove code duplication

### ISSUE THREAD

**Issue by adrianbroher** [freeorion/freeorion#1339]
Currently we duplicate the constructor code used to create objects or use a dedicated Init class.  To avoid this unnecessary code duplication C++11 the concept of delegated constructors.  Delegated constructors allow the developer to call another constructor within the same class within the initialization list to remove said code duplication.  For example:

``` C++
class Foo
{
public:
    Foo() :
        Foo("", 0.0f, nullptr)
    {}

    Foo(const std::string& name, float weight, Foo* ancestor) :
       m_name(name),
       m_weight(m_weight),
       m_ancestor(ancestor)
   {
       if (ancestor && name.empty())
           throw std::invalid_argument("Foo::Foo(): foo instances with ancestor require a name");
   }

private:
    std::string m_name;
    float m_weight;
    Foo* ancestor;
};
```

This reduces possible code duplication in the constructor body, initializer lists and removes the need of `Init` member functions.

Pitfalls:

* When delegating the constructor the delegated constructor must be the only element within the initialization list.

Possible candidates for review were collected by calling:

```bash
for CLASS in $(grep --exclude=stutil.py --exclude-dir=cmake --exclude-dir default -rh '\<class\>' * | sed -e 's/^ *//;/^class/!d;/\\$/d;/Calling Visit/d;/cannot call/d;/streaming operator/d;/MultiEditStyle flags/d;s/: .*//;s/FO_COMMON_API //;s/GG_API //;s/;.*$//;s/ {}\?$//;s/ *$//;s/ \/\/.*//;s/ :$//;s/^class //;s/.*:://;/m_self/d;/m_app/d' | sort -u); do
     (echo ${CLASS} && grep -rn --exclude-dir default " \<${CLASS}(" * | sed '/\<new\>/d;/\<void\>/d') >> ../constructor-candidates ;
done
```

from within the repository root.  The list was reviewed after that and the current result can be viewed via:

https://gist.github.com/adrianbroher/2a6eb71a3fc45896b6a8f7dea39cfbf3

**Comment by memem359** [comment]
Is the status still "help wanted"?
I could start working on that suggested list, at the end of the initial post.

But it would help to know if I'm about to duplicate work. (Annoying to spend the time, and then find that someone else already committed the change just before I was about to.)
Or if you want it handled a certain way. (Submit PR for one file at a time, or a bunch, or wait until almost all are ready?)

**Comment by adrianbroher** [comment]
> Is the status still "help wanted"?
>
> But it would help to know if I'm about to duplicate work. (Annoying to spend the time, and then find that someone else already committed the change just before I was about to.)

Yes, you can still tackle the issue if you want.  As far as I know nobody is working on this issue and nobody has fixed/reviewed the offending lines in detail.

> Or if you want it handled a certain way. (Submit PR for one file at a time, or a bunch, or wait until almost all are ready?)

I personally would prefer a single PR containing all related changes for this Issue with a commit by changed file.  It's hard to suggest a process that works for every PR, but in general it is a good idea to keep commits as small and atomic as possible (e.g. don't mix refactoring, features additions and style changes in a single commit or e.g. when you change a class interface make sure that also all usages of said interfaces are updated in the same commit) and try to ensure that every commit works.

**Comment by adrianbroher** [comment]
Assigning task to @memem359.

**Comment by memem359** [comment]
Short Version:
After review of the code in that list, I recommend that this issue be closed without code changes. Any improvements would be negligible compared to the risk of breaking something. Someone with Wiki write privileges could add comments about delegated contructors to the [Code Standards](http://www.freeorion.org/index.php/Code_Standards) page.

Long version:
Looked over the 99 class/structs that were in the list, excluding the GG code.
91 of them had one or both initializers with 0 or 1 line in the constructor body, so the code would not be streamlined. 1 had no overlap between the code used by the 2 constructor bodies (very different parameters in the argument lists).

For the remaining 7, the overlapping code was already in a separate Init member function.
I think it would be safer to leave things as they are. (Rearranging the deck chairs, so to speak, might introduce more bugs without reducing code duplication. More difficult to find out if that Init member function is only being used during initialization, or also called elsewhere.)

There are likely gains to be made by removing code duplication, but those gains are not within the initializers.

**Comment by adrianbroher** [comment]
> 91 of them had one or both initializers with 0 or 1 line in the constructor body, so the code would not be streamlined.

The duplication doesn't end with the body. The initialization list itself can be duplication enough that needs to be removed.

> For the remaining 7, the overlapping code was already in a separate Init member function. I think it would be safer to leave things as they are.

Init member functions are a pre C++11 crutch.

**Comment by memem359** [comment]
Good point about the list itself.
I'll take another look (after 0.4.7 is released).

Init member functions aren't inherently wrong. The work was already done, and the crutch is in place.
(If it ain't broke, don't fix it.)
If I have to do work in a class to simplify the initialization list, I'll see about reorganize the init member too.
